### PR TITLE
Upgrade to Angular 9

### LINKED ISSUE #11255: angular: Upgrade to v9.0
##### **Overview of the feature request**

Angular v9 is finally out with Ivy support and breaking changes. Check
https://blog.angular.io/version-9-of-angular-now-available-project-ivy-has-arrived-23c97b63cfa3
##### **Related issues or PR**

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by wmarques**
Fix #11255

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/client/templates/angular/webpack/webpack.common.js.ejs [hunk 1]
```diff
@@ -117,6 +122,11 @@ module.exports = (options) => ({
             chunksSortMode: 'manual',
             inject: 'body'
         }),
-        new BaseHrefWebpackPlugin({ baseHref: '/' })
+        new BaseHrefWebpackPlugin({ baseHref: '/' }),
+        new AngularCompilerPlugin({
+            mainPath: utils.root('src/main/webapp/app/app.main.ts'),
```
Comment (ruddell) [inline generators/client/templates/angular/webpack/webpack.common.js.ejs]:
```suggestion
            mainPath: utils.root('<%= MAIN_SRC_DIR %>app/app.main.ts'),
```

**Comment by wmarques** [comment]
I couldn't use TestBed.inject instead of TestBed.get because we use mock classes instead of the real service so the typing is not working.

**Comment by wmarques** [comment]
Thanks @ruddell ! Hope that's good, I tested it on my laptop with some entities... 

Let's see if the CI is ok 🙏

**Comment by pascalgrimaud** [comment]
FYI, the CI failed during the packaging phase here

**Comment by wmarques** [comment]
I need a release after https://github.com/jhipster/ng-jhipster/pull/124 is merged and we should - finally - be good 🎉

**Comment by mshima** [comment]
It’s missing tsconfig-aot.json cleanup.

**Comment by wmarques** [comment]
@mshima No I think I removed it, where do you see it ?

**Comment by mshima** [comment]
> @mshima No I think I removed it, where do you see it ?

I wasn’t clear. The file should be removed from old projects.
https://github.com/jhipster/generator-jhipster/blob/master/generators/cleanup.js

**Comment by postpersonality** [comment]
Hi. I've applied the changes manually to my existing project and I've noticed lots of errors like
```
/src/test/javascript/spec/app/entities/<entity>/<entity>.component.spec.ts is part of the TypeScript compilation but it's unused.
Add only entry points to the 'files' or 'include' properties in your tsconfig.
```
when doing `npm run webpack:build`

And the second thing I've noticed you have `"@angular/common": "9.0.0"` in the package.json while there are updates to the `9.0.3`.

**Comment by wmarques** [comment]
@postpersonality alright i'll check this out, thanks for reaching up

**Comment by postpersonality** [comment]
There is also 
```
npm WARN @ng-bootstrap/ng-bootstrap@5.2.1 requires a peer of @angular/common@^8.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @ng-bootstrap/ng-bootstrap@5.2.1 requires a peer of @angular/core@^8.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @ng-bootstrap/ng-bootstrap@5.2.1 requires a peer of @angular/forms@^8.0.0 but none is installed. You must install peer dependencies yourself.
```
during the `npm install`

I think `@ng-bootstrap/ng-bootstrap` needs to be updated to `6.0.0`

**Comment by wmarques** [comment]
Looks like 9.0.3 is not ready : https://github.com/angular/angular/issues/35709
I saw this issue on my project so I prefer not upgrading to this version.

**Comment by postpersonality** [comment]
> Looks like 9.0.3 is not ready : angular/angular#35709
Thanks for posting this. Got the same problem with the 9.0.3. The 9.0.2 works fine for me.

**Comment by wmarques** [comment]
@postpersonality I don't have the TS warnings. Did you add
`  "files": ["src/main/webapp/app/app.main.ts"]` in `tsconfig.app.json` ?

**Comment by wmarques** [comment]
I think this is ready for merge.
 cc @kaidohallik @vishal423 if you want to check this 😄

**Comment by vishal423** [comment]
@wmarques, I will try to have a quick pass by today evening or tomorrow.

File: generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs [hunk 2]
```diff
@@ -18,4 +18,5 @@
 -%>
 import 'core-js/proposals/reflect-metadata';
 import 'zone.js/dist/zone';
+import '@angular/localize/init';
```
Comment (vishal423) [inline generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs]: as per above comments, check if required

File: generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs [hunk 3]
```diff
@@ -16,6 +16,7 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 -%>
+import './polyfills';
```
Comment (vishal423) [inline generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs]: any specific reason to move to this approach or can we follow cli generated code to add it to `tsconfig.app.json` like 
```"files": ["src/main.ts", "src/polyfills.ts"],```

File: generators/client/templates/angular/tsconfig.json.ejs [hunk 4]
```diff
@@ -18,35 +18,32 @@
...
-    "exclude": [
-        "node_modules"
-    ]
-}
+    "angularCompilerOptions": {
+      "genDir": "<%= AOT_DIR %>",
+      "skipMetadataEmit": true,
+      "fullTemplateTypeCheck": true,
+      "preserveWhitespaces": true,
+      "enableIvy": true
```
Comment (vishal423) [inline generators/client/templates/angular/tsconfig.json.ejs]: I think we can remove this configuration as it's enabled as default.

File: generators/client/templates/angular/tsconfig.json.ejs [hunk 5]
```diff
@@ -18,35 +18,32 @@
...
-        "<%= MAIN_SRC_DIR %>app",
-        "<%= TEST_SRC_DIR %>"
-    ],
-    "exclude": [
-        "node_modules"
-    ]
-}
+    "angularCompilerOptions": {
+      "genDir": "<%= AOT_DIR %>",
+      "skipMetadataEmit": true,
```
Comment (vishal423) [inline generators/client/templates/angular/tsconfig.json.ejs]: do we require this configuration? Per doc, they don't recommend as `true`

**Comment by vishal423** [comment]
> I couldn't use TestBed.inject instead of TestBed.get because we use mock classes instead of the real service so the typing is not working.

In my schematic sample project, I had noticed lint errors/warnings with `TestBed.get`. Can you check that's not the case here?

File: generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs (hunk 3 above)
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs]: I just thought it reduce entry points, I don't see why we put the polyfills apart...

File: generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs (hunk 2 above)
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs]: Yes it is, without invoking init() method, the app cannot start

File: generators/client/templates/angular/package.json.ejs [hunk 6]
```diff
@@ -60,9 +61,9 @@
     "zone.js": "0.10.2"
   },
   "devDependencies": {
-    "@angular/cli": "8.3.20",
-    "@angular/compiler-cli": "8.2.14",
-    "@ngtools/webpack": "8.3.20",
+    "@angular/cli": "9.0.1",
+    "@angular/compiler-cli": "9.0.0",
+    "@ngtools/webpack": "9.0.1",
```
[the remark this answers is not in the document]
Comment (wmarques) [reply generators/client/templates/angular/package.json.ejs]: Yes I'll check this thanks

File: generators/client/templates/angular/package.json.ejs [hunk 7]
```diff
@@ -27,17 +27,18 @@
...
-    "@angular/forms": "8.2.14",
-    "@angular/platform-browser": "8.2.14",
-    "@angular/platform-browser-dynamic": "8.2.14",
-    "@angular/router": "8.2.14",
-    "@fortawesome/angular-fontawesome": "0.5.0",
+    "@angular/common": "9.0.0",
+    "@angular/compiler": "9.0.0",
+    "@angular/core": "9.0.0",
+    "@angular/forms": "9.0.0",
+    "@angular/localize": "^9.0.0",
```
[the remark this answers is not in the document]
Comment (mraible) [reply generators/client/templates/angular/package.json.ejs]: It's required by ng-bootstrap.

File: generators/client/templates/angular/package.json.ejs [hunk 8]
```diff
@@ -27,17 +27,18 @@
...
+    "@angular/platform-browser": "9.0.0",
+    "@angular/platform-browser-dynamic": "9.0.0",
+    "@angular/router": "9.0.0",
```
Comment (mraible) [inline generators/client/templates/angular/package.json.ejs]: Angular 9.0.4 is available https://github.com/angular/angular/releases.

**Comment by mraible** [comment]
I'm assigning a bug bounty to this one because it wasn't that straightforward and its an important feature for JHipster.

File: generators/client/templates/angular/package.json.ejs (hunk 7 above)
[the remark this answers is not in the document]
Comment (vishal423) [reply generators/client/templates/angular/package.json.ejs]: that makes sense

File: generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs [hunk 9]
```diff
@@ -18,4 +18,5 @@
 -%>
 import 'core-js/proposals/reflect-metadata';
```
Comment (vishal423) [inline generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs]: it just came to my mind that we also don't require it any more as we would be using ngtools to build with ivy

File: generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs (hunk 3 above)
Comment (vishal423) [reply generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs]: can you confirm that it still produces a different bundle for polyfills? If not, then, it's a deviation from current behavior and also wouldn't recommend that.

**Comment by pascalgrimaud** [comment]
@mraible : there is already a bounty on the issue, but if you think it's not enough, I'm increasing it and remove the bounty on the PR. No need to add bounty on both issue and PR, it can confuse people :p

**Comment by wmarques** [comment]
Thanks @mraible and @vishal423, should be good now :)

File: generators/client/templates/angular/package.json.ejs [hunk 10]
```diff
@@ -27,14 +27,14 @@
...
-    "@angular/forms": "9.0.0",
-    "@angular/localize": "^9.0.0",
-    "@angular/platform-browser": "9.0.0",
-    "@angular/platform-browser-dynamic": "9.0.0",
-    "@angular/router": "9.0.0",
+    "@angular/common": "9.0.4",
+    "@angular/compiler": "9.0.4",
+    "@angular/core": "9.0.4",
+    "@angular/forms": "9.0.4",
+    "@angular/localize": "^9.0.4",
```
Comment (vishal423) [inline generators/client/templates/angular/package.json.ejs]: can use exact version?

File: generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs (hunk 3 above)
Comment (vishal423) [reply generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs]: @wmarques, can you confirm the behavior?

**Comment by wmarques** [comment]
Only one React build failing 😢

File: generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs (hunk 3 above)
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/app.main.ts.ejs]: It doesn't produces a different bundle but the polyfill would still be loaded so for me that's not really important, polyfills are always loaded, we don't do differential loading like angular CLI does.

File: generators/client/templates/angular/package.json.ejs [hunk 11]
```diff
@@ -146,7 +147,7 @@
     "ts-node": "8.5.4",
     <%_ } _%>
     "tslint": "5.20.1",
```
[the remark this answers is not in the document]
Comment (wmarques) [reply generators/client/templates/angular/package.json.ejs]: done :)

**Comment by wmarques** [comment]
> > I couldn't use TestBed.inject instead of TestBed.get because we use mock classes instead of the real service so the typing is not working.
> 
> In my schematic sample project, I had noticed lint errors/warnings with `TestBed.get`. Can you check that's not the case here?

Didn't notice anything with `npm run lint`

**Comment by wmarques** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/14646

**Pull request merged**